File: algorithms/leetcode/heap/q295_medianFinder.js
/**
 * 295. 数据流的中位数 https://leetcode.cn/problems/find-median-from-data-stream/description/?envType=study-plan-v2&envId=top-100-liked
 * @constructor
 */
const MedianFinder = function() {
    this.min_heap = new MinHeap();
    this.max_heap = new BigHeap();
};

/**
 * @param {number} num
 * @return {void}
 */
MedianFinder.prototype.addNum = function(num) {
    if (this.min_heap.heap.length === this.max_heap.heap.length) {
        if (num > this.max_heap.heap[0]) {
            this.min_heap.insert(num);
        } else {
            this.max_heap.insert(num);
        }
    } else if (this.min_heap.heap.length > this.max_heap.heap.length) {
        if (num < this.min_heap.heap[0]) {
            this.max_heap.insert(num);
        } else {
            this.max_heap.insert(this.min_heap.heap[0]);
            this.min_heap.heap[0] = num;
            this.min_heap.shiftDown(0);
        }
    } else if (this.min_heap.heap.length < this.max_heap.heap.length) {
        if (num > this.max_heap.heap[0]) {
            this.min_heap.insert(num);
        } else {
            this.min_heap.insert(this.max_heap.heap[0]);
            this.max_heap.heap[0] = num;
            this.max_heap.shiftDowm(0);
        }
    }
};

/**
 * @return {number}
 */
MedianFinder.prototype.findMedian = function() {
    if (this.min_heap.heap.length === this.max_heap.heap.length) {
        return (this.min_heap.heap[0] + this.max_heap.heap[0]) / 2;
    } else if (this.min_heap.heap.length > this.max_heap.heap.length) {
        return this.min_heap.heap[0];
    } else {
        return this.max_heap.heap[0];
    }
};

// 小根堆
class MinHeap {
    heap = [];

    swap(index1, index2) {
        let temp = this.heap[index1];
        this.heap[index1] = this.heap[index2];
        this.heap[index2] = temp;
    }

    // 插入一个元素
    insert(value) {
        this.heap.push(value);
        this.shiftUp(this.heap.length - 1);
    }

    shiftUp(index) {
        let parentNodeIndex = (index - 1) >>> 1;
        if (index !== 0 && this.heap[index] < this.heap[parentNodeIndex]) {
            this.swap(parentNodeIndex, index);
            this.shiftUp(parentNodeIndex);
        }
    }

    shiftDown(index) {
        let leftNodeIndex = (index + 1) * 2 - 1;
        let rightNodeIndex = (index + 1) * 2;

        if (leftNodeIndex < this.heap.length && this.heap[leftNodeIndex] < this.heap[index]) {
            this.swap(index, leftNodeIndex);
            this.shiftDown(leftNodeIndex);
        }
        if (rightNodeIndex < this.heap.length && this.heap[rightNodeIndex] < this.heap[index]) {
            this.swap(index, rightNodeIndex);
            this.shiftDown(rightNodeIndex);
        }
    }
}

// 大根堆
class BigHeap {
    heap = [];

    swap(index1, index2) {
        let temp = this.heap[index1];
        this.heap[index1] = this.heap[index2];
        this.heap[index2] = temp;
    }

    insert(value) {
        this.heap.push(value);
        this.shiftUp(this.heap.length - 1);
    }

    shiftUp(index) {
        let parentNodeIndex = (index - 1) >>> 1;
        if (index !== 0 && this.heap[index] > this.heap[parentNodeIndex]) {
            this.swap(index, parentNodeIndex);
            this.shiftUp(parentNodeIndex);
        }
    }

    shiftDowm(index) {
        let leftNodeIndex = (index + 1) * 2 - 1;
        let rightNodeIndex = (index + 1) * 2;

        // 左子树节点值 > 根节点值
        if (leftNodeIndex < this.heap.length && this.heap[leftNodeIndex] > this.heap[index]) {
            this.swap(leftNodeIndex, index);
            this.shiftDowm(leftNodeIndex);
        }
        if (rightNodeIndex < this.heap.length && this.heap[rightNodeIndex] > this.heap[index]) {
            this.swap(rightNodeIndex, index);
            this.shiftDowm(rightNodeIndex);
        }
    }
}

/**
 * Your MedianFinder object will be instantiated and called as such:
 * var obj = new MedianFinder()
 * obj.addNum(num)
 * var param_2 = obj.findMedian()
 */

const heap = new MedianFinder();
heap.addNum(1);
heap.addNum(2);
console.log(heap.findMedian());
heap.addNum(3);
console.log(heap.findMedian());

File: algorithms/leetcode/heap/q295_medianFinder.java
class MedianFinder {
    // 大顶堆存放前 k/2 个元素，堆顶元素是前半段数据的结束位置
    private PriorityQueue<Integer> maxHeap;
    // 小顶堆存放后 k/2 个元素，堆顶元素为后半段数据的开始位置
    private PriorityQueue<Integer> minHeap;

    public MedianFinder() {
        minHeap = new PriorityQueue<>();
        maxHeap = new PriorityQueue<>(Comparator.reverseOrder());
    }

    public void addNum(int num) {
        if (maxHeap.size() == minHeap.size()) {
            maxHeap.offer(num);
            minHeap.offer(maxHeap.poll());
        } else {
            minHeap.offer(num);
            maxHeap.offer(minHeap.poll());
        }
    }

    // 当数据个数为奇数时，大顶堆的元素个数大于小顶堆元素个数取，大顶堆的堆顶元素作为中位数，否则反之
    // 当数据个数为偶数时，取大顶堆与小顶堆的堆顶之和的平均值作为中位数
    public double findMedian() {
        if (minHeap.size() == maxHeap.size()) {
            return (minHeap.element() + maxHeap.element()) / 2.0;
        }
        return minHeap.size() < maxHeap.size() ? maxHeap.element() : minHeap.element();
    }
}

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder obj = new MedianFinder();
 * obj.addNum(num);
 * double param_2 = obj.findMedian();
 */